File: 백준/Silver/1012. 유기농 배추/유기농 배추.py
import sys
input = sys.stdin.readline
from collections import deque

def bfs(s): # 인접한 곳의 값을 전부 0으로 만들기
    Q = deque()
    Q.append(s)
    while Q:
        row, col = Q.popleft()
        if arr[row][col]: # 아직 방문을 안 했다면
            arr[row][col] = 0
            for dr, dc in [(1, 0), (0, 1), (-1, 0), (0, -1)]: # 아래 오룬 위 왼
                newR = row + dr
                newC = col + dc
                if 0<= newR < N and 0<= newC < M and arr[newR][newC]: # 범위에 벗어나 있지 않으며, 배추가 있으면
                    Q.append((newR, newC))

T = int(input()) # 테스트 케이스

for _ in range(T):
    M, N, K = map(int, input().split()) # 열(가로) 행(세로) 배추 개수
    arr = [[0]*M for _ in range(N)]
    # print(arr)
    for _ in range(K):
        X, Y = map(int, input().split())
        arr[Y][X] = 1
    # print(arr)

    cnt = 0 # 필요한 지렁이의 수 카운트

    for r in range(N):
        for c in range(M):
            if arr[r][c] == 1:
                cnt += 1
                bfs((r, c))

    print(cnt)

File: 백준/Silver/1012. 유기농 배추/유기농 배추.java
import java.io.*;
import java.util.*;

public class Main {
  static int N;
  static int M;
  static int[][] arr;
  static boolean[][] visited;

  public static void bfs(int stR, int stC) {
    Deque<int[]> dq = new ArrayDeque();
    dq.offerLast(new int[] {stR, stC});

    int[][] dir = new int[][] {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};
    while (!dq.isEmpty()) {
      int[] cur = dq.pollFirst();
      for (int i = 0; i < 4; i++) {
        int newR = cur[0] + dir[i][0];
        int newC = cur[1] + dir[i][1];
        if (newR < 0 || newR >= N || newC < 0 || newC >= M) continue;
        if (arr[newR][newC] == 1 && !visited[newR][newC]) {
          visited[newR][newC] = true;
          dq.offerLast(new int[] {newR, newC});
        }
      }
    }
  }

  public static void main(String[] args) throws Exception {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    int T = Integer.parseInt(br.readLine());
    StringTokenizer st;
    // 테스트 케이스만큼 반복
    StringBuilder sb = new StringBuilder();
    while (T-- > 0) {
      st = new StringTokenizer(br.readLine());
      N = Integer.parseInt(st.nextToken()); // 행
      M = Integer.parseInt(st.nextToken()); // 열
      int K = Integer.parseInt(st.nextToken()); // 배추가 심어져 있는 땅의 수
      arr = new int[N][M]; // 배추밭
      visited = new boolean[N][M];

      int row;
      int col;
      while (K-- > 0) {
        st = new StringTokenizer(br.readLine());
        row = Integer.parseInt(st.nextToken());
        col = Integer.parseInt(st.nextToken());
        arr[row][col] = 1;
      }

      int cnt = 0;
      for (int r = 0; r < N; r++) {
        for (int c = 0; c < M; c++) {
          if (arr[r][c] == 1 && !visited[r][c]) {
            visited[r][c] = true;
            cnt++;
            bfs(r, c);
          }
        }
      }
      sb.append(cnt).append('\n');
    }
    // 출력
    System.out.println(sb);
  }
}
